### PR TITLE
fix(material/tabs): ensure the ink bar realigns when the tab header changes dimensions

### LINKED ISSUE #3048: [Tabs] Ink bar is not correctly aligned on init
#### Bug, feature request, or proposal:
Bug (Some different than https://github.com/angular/material2/issues/3044)

#### What is the expected behavior?
Highlighter [md-ink-bar] should invalidate position on init app

#### What is the current behavior?
Misaligned [md-ink-bar]. Please see attached image.

#### What are the steps to reproduce?
_Case 1_
1. Open demo application from material2 
2. Navigate to tabs (Especially for _Tab Nav Bar_ with routing example)
3. Apply **md-stretch-tabs** directive for <nav> and apply  styles for tab items 
`[md-stretch-tabs] [md-tab-link] {
  flex-basis: 0;
  flex-grow: 1;
}`
4. Try to expand / collapse application sidebar.

_Case 2_ (or _just apply attached git patch._ )
1. Open demo application from material2 
2. Setup  <md-sidenav **mode="side" opened="true"**>  for root app container.
3. Navigate to tabs (Especially for _Tab Nav Bar_ with routing example)
4. Apply **md-stretch-tabs** directive for <nav> and apply  styles for tab items 
`[md-stretch-tabs] [md-tab-link] {
  flex-basis: 0;
  flex-grow: 1;
}`
5. Select not first tab
6. Refresh page

Providing a Plunker (or similar) is the *best* way to get the team to see your issue.
Patch: [MD_TabBar_issue.patch.zip](https://github.com/angular/material2/files/767999/MD_TabBar_issue.patch.zip)

#### Which versions of Angular, Material, OS, browsers are affected?
Based on  up to date sources of [material2](material2)

#### Is there anything else we should know?
image

### LINKED ISSUE #6130: [Tabs] Ink bar does not update when container resizes
#### Bug, feature request, or proposal:
Bug :
Add stretched tabs in a sidenav container with side mode.
When the sidenav is resized with dynamic content (ie sidenav labels are translated), the md-ink-bar is not resized
That happens too with md-tab-nav-bar and full width tabs

#### What is the expected behavior?
md-ink-bar should be resized

#### What is the current behavior?
tabs are well resized but not md-ink-bar

#### What are the steps to reproduce?

plunkr :
https://plnkr.co/edit/q80nul7KYS8sJ0e8vwwi?p=preview
stackblitz :
https://stackblitz.com/edit/angular-material2-issue-6efd2m?file=main.ts

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
ng 5.1.1
material 5.0.1
chrome 61

### PR DESCRIPTION
**Pull request by hilts-vaughan**
This should fix #6130 and also #4648?. It uses `ResizeObserver` to ensure we get notifications about the dimensions of the tabs when they change. The layout change will fire synchronously. 

It may also fix #3048.

### CODE REVIEW

File: src/material/tabs/paginated-tab-header.ts
Comment (crisbeto) [inline src/material/tabs/paginated-tab-header.ts:256]: This needs to be feature detected since it could run in a server environment.

File: src/material/tabs/paginated-tab-header.ts
Comment (crisbeto) [inline src/material/tabs/paginated-tab-header.ts:265]: Rather than observing each individual element, we could observe the container that they're in.

File: src/material/tabs/paginated-tab-header.ts
Comment (hilts-vaughan) [reply src/material/tabs/paginated-tab-header.ts:256]: Does Angular have their own mechanism for this or can I just do a `window.ResizeObserver` check?

File: src/material/tabs/paginated-tab-header.ts
Comment (hilts-vaughan) [reply src/material/tabs/paginated-tab-header.ts:265]: I think that would be observing this which would trigger more changes than needed I think since the container changing sizes doesn't actually change the contents of their boxes.

What do you think?

File: src/material/tabs/paginated-tab-header.ts
Comment (crisbeto) [reply src/material/tabs/paginated-tab-header.ts:256]: You can get it from `this._platform.isBrowser`, although I think I'd prefer to use `typeof ResizeObserver === 'function'` since it would avoid errors on browsers that don't support `ResizeObserver`.

File: src/material/tabs/paginated-tab-header.ts
Comment (crisbeto) [reply src/material/tabs/paginated-tab-header.ts:265]: I think that in most cases changing the size of the container will change the position of the ink bar. It's also worth it for the reduced complexity.

Comment (crisbeto) [inline src/material/tabs/paginated-tab-header.ts:265]: We should run the observer outside of the `NgZone`.

File: src/material/tabs/tab-header.spec.ts
Comment (crisbeto) [inline src/material/tabs/tab-header.spec.ts:648]: I suspect that this change is going to cause flakiness, because we can't assume that the resize observer will fire immediately.

File: src/material/tabs/paginated-tab-header.ts
Comment (hilts-vaughan) [reply src/material/tabs/paginated-tab-header.ts:265]: The container itself is full width in a lot of cases; I just checked. The tabs are then centered within it (at least in the demo app) so this won't actually work. I'm happy to come up with something else if you think this would be a problem but ResizeObserver already batches all the changes into a single call so it's probably not a problem for performance either. 

I do see that ResizeObserver will remove old calls to the `observe` when duplicated though, so I removed the set checking. Does that help?

Comment (hilts-vaughan) [reply src/material/tabs/paginated-tab-header.ts:265]: Done.

File: src/material/tabs/tab-header.spec.ts
Comment (hilts-vaughan) [reply src/material/tabs/tab-header.spec.ts:648]: Could you explain? By definition, the ResizeObserver has to fire at the next frame before the paint and this is intentional and specced to avoid frame jank. I can deflake the test -- I just ran 50 times locally as well and never had a single flake.

File: src/material/tabs/paginated-tab-header.ts
Comment (crisbeto) [reply src/material/tabs/paginated-tab-header.ts:271]: The call above should be `runOutsideAngular`, not `run`.

File: src/material/tabs/tab-header.spec.ts
Comment (crisbeto) [reply src/material/tabs/tab-header.spec.ts:648]: MDN has the following clarification:

> ResizeObserver avoids infinite callback loops and cyclic dependencies that are often created when resizing via a callback function. It does this by only processing elements deeper in the DOM in subsequent frames.

The way I'm reading this is that it's basically being invoked in a `requestAnimationFrame` which is asynchronous, whereas this test is synchronous.

File: src/material/tabs/paginated-tab-header.ts
Comment (hilts-vaughan) [reply src/material/tabs/paginated-tab-header.ts:271]: Done. Thanks for the tip.

File: src/material/tabs/tab-header.spec.ts
Comment (hilts-vaughan) [reply src/material/tabs/tab-header.spec.ts:648]: It's scheduled later but the change by definition has to be done before the paint is finished so that the change is reflected. I don't understand how the zone in the test is run to know if the change is reflected in the DOM but the test is not flakey on current runs. 

Did you have a preference on a way to test this that you would prefer? I'm open to suggestions. I noticed that other cases in the codebase that use RO don't even have tests in their PR that run as part of the changes, perhaps for a similar reason.

File: src/material/tabs/paginated-tab-header.ts
Comment (crisbeto) [inline src/material/tabs/paginated-tab-header.ts:270]: This still hasn't been changed to `runOutSizeAngular`.

File: src/material/tabs/paginated-tab-header.ts
Comment (crisbeto) [inline src/material/tabs/paginated-tab-header.ts:262]: I don't think that we should be stubbing out an observable. You can use the `EMPTY` observable from rxjs instead.

**Review by crisbeto — COMMENTED** [review]
It seems like the CI didn't run for some reason 😕

File: src/material/tabs/paginated-tab-header.ts
Comment (hilts-vaughan) [reply src/material/tabs/paginated-tab-header.ts:270]: Sorry, I forgot to push.

File: src/material/tabs/paginated-tab-header.ts
Comment (hilts-vaughan) [reply src/material/tabs/paginated-tab-header.ts:262]: We're not. We're stubbing out the RO function. I could also return `null` and check for `null` down below. I don't think EMPTY works.

Comment (crisbeto) [reply src/material/tabs/paginated-tab-header.ts:262]: Hmm, I see. I find this side-effect-ey use of `map` confusing. Here's how I would've implemented it:

1. Keep the `merge(dirChange, resize, this._items.changes)` above as is.
2. Have a utility that looks like this:
```
private _itemsResized(): Observable<void> {
  if (typeof ResizeObserver !== 'function') {
    return EMPTY;
  }

  return this._items.changes.pipe(
    startWith(this._items),
    switchMap(
      (items: QueryList<MatPaginatedTabHeaderItem>) =>
        new Observable((observer: Observer<void>) =>
          this._ngZone.runOutsideAngular(() => {
            const resizeObserver = new ResizeObserver(() => observer.next());
            items.forEach(item => resizeObserver.observe(item.elementRef.nativeElement));
            return () => resizeObserver.disconnect();
          }),
        ),
    ),
    // Skip the first emit since the resize observer emits when an item is observed.
    skip(1),
  );
}
```
3. Change the `merge` call above to `merge(dirChange, resize, this._items.changes, this._itemsResized())`.


This approach fixes the following issues:
1. You generally don't want side effects from your `map` operator.
2. The `ResizeObserver` wasn't being disconnected when the subscription is dropped.
3. It avoids the extra call when the elements are observed for the first time.
4. The `ResizeObserver` itself wasn't being created outside the `NgZone`.

Note that this is slightly more wasteful when it comes to recreating resize observers, but I think that it's not a big deal in this case since tabs don't tend to change very often. If it becomes a problem, we can start keeping track of which items are being observed.

Comment (crisbeto) [reply src/material/tabs/paginated-tab-header.ts]: You can reintroduce the first render by removing the `skip(1)`. I added it, because the first emit seems redundant to me. It fires when the elements are observed for the first time, but the initial render should already be correct since we do it right after the elements are inserted.

Comment (hilts-vaughan) [reply src/material/tabs/paginated-tab-header.ts]: Removing `skip` will fire it on the RO layout which causes the rest to fail because as it turns out, the `fakeAsyncZone` doesn't actually capture the `ResizeObserver` tick (even though it technically runs before the `paint`, it's not before the call stack is cleared). This explains why that test passed and was not flaky.

We could remove the test and not bother or perhaps try to a stability check hack to fix that if you don't think the first synchronous emission is needed.

Comment (crisbeto) [reply src/material/tabs/paginated-tab-header.ts]: Okay, that explains why the test was passing. I'd rather avoid the test than have it fire on init, because it has the potential to cause layout thrashing.

Comment (hilts-vaughan) [reply src/material/tabs/paginated-tab-header.ts]: OK, I removed the test for now to keep it simple especially since MDC is coming "soon". I talked with some internal folks and these links should be OK to share:

Before: https://screencast.googleplex.com/cast/NTA3MjkzNjU0MjE0MjQ2NHw0NTU1YzBmNi1lOA
After: https://screencast.googleplex.com/cast/NjI3MjI0NjU2Njg3OTIzMnw0MjBiMDcxZS0zNQ

This shows the stuff working in lieu of some tests.

Comment (crisbeto) [reply src/material/tabs/paginated-tab-header.ts]: Shouldn't we get rid of the `startWith(undefined)` as well?

Comment (hilts-vaughan) [reply src/material/tabs/paginated-tab-header.ts]: It should be gone. I hadn't realize that pushing would immediately send it for re-review, unlike some other tools. :)

**Comment by crisbeto** [comment]
So the change LGTM, but for some reason the CI didn't run. Can you try cherry-picking the commit to a new branch and opening another PR?

**Comment by hilts-vaughan** [comment]
The CI system says it's missing some labels. Do I need to add some labels on the new PR? Or this one perhaps?

**Comment by crisbeto** [comment]
The label checks are separate. Usually it queues up another 11 jobs that don't show up here for some reason. This has happened for other people and CircleCI haven't been able to figure out why. There was a theory that it might be because you aren't watching the fork that you're opening the PR from.

**Comment by hilts-vaughan** [comment]
Opened https://github.com/angular/components/pull/24932. I'll adjust the description etc if the CI runs.

**Comment by hilts-vaughan** [comment]
(It also tells me that the CI workflow is pending approval and links me to https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks -- claiming a maintainer has to approve my CI run on this PR)

**Comment by crisbeto** [comment]
Did you try clicking the "Watch" button on your fork? Also the "Approve and run" button is only for deploying the dev app, it doesn't affect any of the other checks. You can check what other PRs look like, e.g. https://github.com/angular/components/pull/24931

**Comment by hilts-vaughan** [comment]
I swapped the watch to "All Activity" just now.

**Comment by crisbeto** [comment]
Then I really don't know why it didn't run 😕. Anyway, I ran it locally and the only problem is this lint failure:

```
src/material/tabs/paginated-tab-header.ts:258:3
ERROR: 258:3  jsdoc-format  jsdoc is not formatted correctly on this line
```

**Comment by hilts-vaughan** [comment]
I tried fix it locally. Want me to run something else? Should I abandon https://github.com/angular/components/pull/24932?

**Comment by crisbeto** [comment]
Yeah, you can abandon the other one.

**Review by crisbeto — APPROVED** [review]
Caretaker note: for some reason the CI didn't run on this PR. I've verified that everything passes on my machine.

**Pull request merged**
